### ISSUE TITLE
Virtual camera fails to start

### ISSUE THREAD

**Issue by topongo** [obsproject/obs-studio#3929]
## Platform

Operating system and version: Arch Linux - Rolling
OBS Studio version: 26.1.0.r9

## Expected Behavior

Clicking on "Start Virtual Camera" button should start the virtual cam flawlessly

## Current Behavior

Pressing the button does not create any effect.

## Steps to Reproduce

1. Install obs-studio-git from AUR
2. Remove the old CatxFish/obs-v4l2sink plugin
3. Install v4l2loopback-dkms
4. Start OBS and click "Start Virtual Camera" button.

## Additional information

The log says "warning: Failed to start virtual camera" when clicking on the button.
I've installed droidcam alongside OBS, and it currently uses the /dev/video0 device. I thought maybe obs is trying to use that device and it conflicts with droidcam, but disabling (not removing) droidcam doesn't solve the issue.

**Comment by pardomi** [comment]
I am experiencing the same issue.

**Platform**
Operating system and version: Kubuntu 20.10
OBS Studio version: 26.1

**Expected Behavior**
Clicking on "Start Virtual Camera" button should start the virtual cam flawlessly

**Current Behavior**
Pressing the button does not create any effect.  (button text remains "Start Virtual Camera"

**Steps to Reproduce**
    Install obs-studio 26.1 from ppa:obsproject/obs-studio
    Install v4l2loopback-dkms (previously installed)
    Start OBS and click "Start Virtual Camera" button.

**Additional information**

- The log says "warning: Failed to start virtual camera" when clicking on the button.
- I do not have droidcam or any other camera manipulation software installed (as far as I know)
- I had a previous version of OBS Studio installed (from Ubuntu 20.10 repository - uninstalled before installing 26.1 from PPA).

**Log File Contents**
03:19:54 PM.696: CPU Name: Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
03:19:54 PM.696: CPU Speed: 1830.528MHz
03:19:54 PM.696: Physical Cores: 2, Logical Cores: 4
03:19:54 PM.696: Physical Memory: 11850MB Total, 2914MB Free
03:19:54 PM.696: Kernel Version: Linux 5.8.0-33-generic
03:19:54 PM.696: Distribution: "Ubuntu" "20.10"
03:19:54 PM.696: Session Type: x11
03:19:54 PM.697: Window System: X11.0, Vendor: The X.Org Foundation, Version: 1.20.9
03:19:54 PM.697: Portable mode: false
03:19:54 PM.754: OBS 26.1.0 (linux)
03:19:54 PM.754: ---------------------------------
03:19:54 PM.841: ---------------------------------
03:19:54 PM.841: audio settings reset:
03:19:54 PM.841: 	samples per sec: 48000
03:19:54 PM.841: 	speakers:        2
03:19:54 PM.850: ---------------------------------
03:19:54 PM.850: Initializing OpenGL...
03:19:54 PM.903: Loading up OpenGL on adapter Intel Mesa Intel(R) HD Graphics 620 (KBL GT2)
03:19:54 PM.903: OpenGL loaded successfully, version 4.6 (Core Profile) Mesa 20.2.1, shading language 4.60
03:19:54 PM.919: ---------------------------------
03:19:54 PM.919: video settings reset:
03:19:54 PM.919: 	base resolution:   1600x900
03:19:54 PM.919: 	output resolution: 1280x720
03:19:54 PM.919: 	downscale filter:  Bicubic
03:19:54 PM.919: 	fps:               30/1
03:19:54 PM.919: 	format:            NV12
03:19:54 PM.919: 	YUV mode:          709/Partial
03:19:54 PM.919: NV12 texture support not available
03:19:54 PM.921: Audio monitoring device:
03:19:54 PM.921: 	name: Default
03:19:54 PM.921: 	id: default
03:19:54 PM.921: ---------------------------------
03:19:54 PM.923: Failed to load 'en-US' text for module: 'decklink-captions.so'
03:19:54 PM.925: Failed to load 'en-US' text for module: 'decklink-ouput-ui.so'
03:19:55 PM.030: A DeckLink iterator could not be created.  The DeckLink drivers may not be installed
03:19:55 PM.030: No blackmagic support
03:19:55 PM.039: [obs-browser]: Version 2.9.1
03:19:55 PM.039: [obs-browser]: CEF Version 76.1.13+gf19c584+chromium-76.0.3809.132
03:19:55 PM.070: os_dlopen(libnvidia-encode.so.1->libnvidia-encode.so.1): libnvidia-encode.so.1: cannot open shared object file: No such file or directory
03:19:55 PM.070: 
03:19:55 PM.070: FFMPEG VAAPI supported
03:19:55 PM.093: VLC found, VLC video source enabled
03:19:55 PM.094: ---------------------------------
03:19:55 PM.094:   Loaded Modules:
03:19:55 PM.094:     vlc-video.so
03:19:55 PM.094:     text-freetype2.so
03:19:55 PM.094:     rtmp-services.so
03:19:55 PM.094:     obs-x264.so
03:19:55 PM.094:     obs-vst.so
03:19:55 PM.094:     obs-transitions.so
03:19:55 PM.094:     obs-outputs.so
03:19:55 PM.094:     obs-libfdk.so
03:19:55 PM.094:     obs-filters.so
03:19:55 PM.094:     obs-ffmpeg.so
03:19:55 PM.094:     obs-browser.so
03:19:55 PM.094:     linux-v4l2.so
03:19:55 PM.094:     linux-pulseaudio.so
03:19:55 PM.094:     linux-jack.so
03:19:55 PM.094:     linux-decklink.so
03:19:55 PM.094:     linux-capture.so
03:19:55 PM.094:     linux-alsa.so
03:19:55 PM.094:     image-source.so
03:19:55 PM.094:     frontend-tools.so
03:19:55 PM.094:     decklink-ouput-ui.so
03:19:55 PM.094:     decklink-captions.so
03:19:55 PM.094: ---------------------------------
03:19:55 PM.094: os_dlopen(../obs-plugins/obs-browser->../obs-plugins/obs-browser.so): ../obs-plugins/obs-browser.so: cannot open shared object file: No such file or directory
03:19:55 PM.094: 
03:19:55 PM.094: ==== Startup complete ===============================================
03:19:55 PM.097: All scene data cleared
03:19:55 PM.097: ------------------------------------------------
03:19:55 PM.106: pulse-input: Server name: 'pulseaudio 13.99.2'
03:19:55 PM.107: pulse-input: Audio format: s16le, 48000 Hz, 2 channels
03:19:55 PM.107: pulse-input: Started recording from 'alsa_output.pci-0000_00_1f.3.analog-stereo.monitor'
03:19:55 PM.107: [Loaded global audio device]: 'Desktop Audio'
03:19:55 PM.110: pulse-input: Server name: 'pulseaudio 13.99.2'
03:19:55 PM.111: pulse-input: Audio format: s16le, 48000 Hz, 2 channels
03:19:55 PM.111: pulse-input: Started recording from 'alsa_input.pci-0000_00_1f.3.analog-stereo'
03:19:55 PM.111: [Loaded global audio device]: 'Mic/Aux'
03:19:55 PM.111: v4l2-input: Start capture from /dev/video0
03:19:55 PM.226: v4l2-input: Input: 0
03:19:55 PM.234: v4l2-input: Resolution: 640x480
03:19:55 PM.234: v4l2-input: Pixelformat: YUYV
03:19:55 PM.234: v4l2-input: Linesize: 1280 Bytes
03:19:55 PM.234: v4l2-input: Framerate: 30.00 fps
03:19:55 PM.235: Switched to scene 'Scene'
03:19:55 PM.235: ------------------------------------------------
03:19:55 PM.235: Loaded scenes:
03:19:55 PM.235: - scene 'Scene':
03:19:55 PM.235:     - source: 'Video Capture Device (V4L2)' (v4l2_input)
03:19:55 PM.235: ------------------------------------------------
03:19:55 PM.665: adding 42 milliseconds of audio buffering, total audio buffering is now 42 milliseconds (source: Mic/Aux)
03:19:55 PM.665: 
03:20:01 PM.687: Failed to start virtual camera
03:20:07 PM.038: ==== Shutting down ==================================================
03:20:07 PM.056: v4l2-input: Stopped capture after 354 frames
03:20:07 PM.098: pulse-input: Stopped recording from 'alsa_output.pci-0000_00_1f.3.analog-stereo.monitor'
03:20:07 PM.098: pulse-input: Got 2073 packets with 487220 frames
03:20:07 PM.098: pulse-input: Stopped recording from 'alsa_input.pci-0000_00_1f.3.analog-stereo'
03:20:07 PM.098: pulse-input: Got 1220 packets with 574616 frames
03:20:07 PM.100: All scene data cleared
03:20:07 PM.100: ------------------------------------------------
03:20:07 PM.232: [Scripting] Total detached callbacks: 0
03:20:07 PM.243: Freeing OBS context data
03:20:07 PM.259: == Profiler Results =============================
03:20:07 PM.259: run_program_init: 699.53 ms
03:20:07 PM.259:  ┣OBSApp::AppInit: 2.373 ms
03:20:07 PM.259:  ┃ ┗OBSApp::InitLocale: 1.074 ms
03:20:07 PM.259:  ┗OBSApp::OBSInit: 622.338 ms
03:20:07 PM.259:    ┣obs_startup: 1.475 ms
03:20:07 PM.260:    ┗OBSBasic::OBSInit: 477.28 ms
03:20:07 PM.260:      ┣OBSBasic::InitBasicConfig: 0.088 ms
03:20:07 PM.260:      ┣OBSBasic::ResetAudio: 0.127 ms
03:20:07 PM.260:      ┣OBSBasic::ResetVideo: 79.832 ms
03:20:07 PM.260:      ┣OBSBasic::InitOBSCallbacks: 0.004 ms
03:20:07 PM.260:      ┣OBSBasic::InitHotkeys: 0.043 ms
03:20:07 PM.260:      ┣obs_load_all_modules: 172.504 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(decklink-captions.so): 0.032 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(decklink-ouput-ui.so): 1.016 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(frontend-tools.so): 74.543 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(image-source.so): 0.006 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(linux-alsa.so): 0.003 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(linux-capture.so): 0.303 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(linux-decklink.so): 0.188 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(linux-jack.so): 0.001 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(linux-pulseaudio.so): 0.001 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(linux-v4l2.so): 2.458 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-browser.so): 0.062 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-ffmpeg.so): 0.299 ms
03:20:07 PM.260:      ┃ ┃ ┗nvenc_check: 0.256 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-filters.so): 0.018 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-libfdk.so): 0.002 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-outputs.so): 0.004 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-transitions.so): 0.009 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-vst.so): 0.006 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(obs-x264.so): 0.004 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(rtmp-services.so): 0.635 ms
03:20:07 PM.260:      ┃ ┣obs_init_module(text-freetype2.so): 0.021 ms
03:20:07 PM.260:      ┃ ┗obs_init_module(vlc-video.so): 2.482 ms
03:20:07 PM.260:      ┣OBSBasic::ResetOutputs: 0.144 ms
03:20:07 PM.260:      ┣OBSBasic::CreateHotkeys: 0.03 ms
03:20:07 PM.261:      ┣OBSBasic::InitService: 1.079 ms
03:20:07 PM.261:      ┣OBSBasic::InitPrimitives: 0.197 ms
03:20:07 PM.261:      ┗OBSBasic::Load: 140.113 ms
03:20:07 PM.261: obs_hotkey_thread(25 ms): min=0.086 ms, median=0.31 ms, max=70.343 ms, 99th percentile=6.86 ms, 99.5885% below 25 ms
03:20:07 PM.261: audio_thread(Audio): min=0 ms, median=0.079 ms, max=2.966 ms, 99th percentile=0.525 ms
03:20:07 PM.261: obs_graphics_thread(33.3333 ms): min=0.017 ms, median=1.388 ms, max=113.687 ms, 99th percentile=16.17 ms, 99.7275% below 33.333 ms
03:20:07 PM.261:  ┣tick_sources: min=0 ms, median=0.012 ms, max=113.6 ms, 99th percentile=0.04 ms
03:20:07 PM.261:  ┣output_frame: min=0.008 ms, median=0.558 ms, max=1.6 ms, 99th percentile=1.525 ms
03:20:07 PM.261:  ┃ ┗gs_context(video->graphics): min=0.007 ms, median=0.558 ms, max=1.598 ms, 99th percentile=1.525 ms
03:20:07 PM.261:  ┃   ┣render_video: min=0.004 ms, median=0.439 ms, max=1.348 ms, 99th percentile=1.312 ms
03:20:07 PM.261:  ┃   ┃ ┗render_main_texture: min=0.004 ms, median=0.433 ms, max=1.345 ms, 99th percentile=1.275 ms
03:20:07 PM.261:  ┃   ┗gs_flush: min=0 ms, median=0.08 ms, max=0.408 ms, 99th percentile=0.366 ms
03:20:07 PM.261:  ┗render_displays: min=0.002 ms, median=0.672 ms, max=18.786 ms, 99th percentile=15.498 ms
03:20:07 PM.261: =================================================
03:20:07 PM.261: == Profiler Time Between Calls ==================
03:20:07 PM.261: obs_hotkey_thread(25 ms): min=25.157 ms, median=25.414 ms, max=95.456 ms, 64.1237% within ±2% of 25 ms (0% lower, 35.8763% higher)
03:20:07 PM.261: obs_graphics_thread(33.3333 ms): min=19.609 ms, median=33.332 ms, max=113.69 ms, 98.9071% within ±2% of 33.333 ms (0.546448% lower, 0.546448% higher)
03:20:07 PM.261: =================================================
03:20:07 PM.275: Number of memory leaks: 0

**Comment by pardomi** [comment]
It seems the issue was that "modprobe v4l2loopback" assigned /dev/video99 to the virtual camera.  Removing v4l2loopback and then manually assigning the next available number (in my case /dev/video2) resulting in the virtual camera working.

I wonder if there is a way to force OBS to use /dev/video99.

**Comment by topongo** [comment]
> It seems the issue was that "modprobe v4l2loopback" assigned /dev/video99 to the virtual camera.  Removing v4l2loopback and then manually assigning the next available number (in my case /dev/video2) resulting in the virtual camera working.
> 
> I wonder if there is a way to force OBS to use /dev/video99.

You mentioned a "modprobe v4l2loopback" but you actually never mentioned it, what do you mean?

**Comment by pardomi** [comment]
I forgot that I had to run "modprobe v4l2loopback" in order for the "Start Virtual Camera" button to appear.  Or at least, the button did not appear until after I ran that command.  I found I had to run that command when experimenting with uninstalling and reinstalling v4l2loopback-dkms as well as OBS.

**Comment by topongo** [comment]
> I forgot that I had to run "modprobe v4l2loopback" in order for the "Start Virtual Camera" button to appear. Or at least, the button did not appear until after I ran that command. I found I had to run that command when experimenting with uninstalling and reinstalling v4l2loopback-dkms as well as OBS.

I see. I've noticed something else: maybe it can be a issue with the propting for elevated rigths, 'cause without root access obs has no way to configure a new v4l2loopback device. So my guess is that obs can't prompt you for elevation, then it fails to register a new virtual camera and consequently fails to start it.

**Comment by lbriais** [comment]
I have "_almost_" the same issue...

Although for me OBS reports that the virtual camera is actually started:

```
[...]
18:04:23.524: Virtual camera started
18:04:23.531: ==== Virtual Camera Start ==========================================
18:05:43.722: Output 'virtualcam_output': stopping
18:05:43.722: Output 'virtualcam_output': Total frames output: 2404
18:05:43.722: Output 'virtualcam_output': Total drawn frames: 2406
18:05:43.722: ==== Virtual Camera Stop ===========================================
18:05:43.724: Virtual camera stopped
18:05:45.202: ==== Shutting down ====q==============================================
[...]
```

The point is I have multiple "virtual" cameras already:

```
$ v4l2-ctl --list-devices      
Virtual akvcam (out) (platform:akvcam-0):
	/dev/video0

Virtual akvcam (platform:akvcam-1):
	/dev/video1

Virtual Camera (platform:v4l2loopback-000):
	/dev/video3

Droidcam (platform:v4l2loopback_dc-000):
	/dev/video2

UVC Camera (046d:0809) (usb-0000:00:14.0-2):
	/dev/video4
```
* The two first are created by the [akvcam module](https://github.com/webcamoid/akvcam) that is used by [webcamoid](https://webcamoid.github.io/).
* The third one is the one that OBS should use (it is not named "OBS Virtual Camera" because I created it in the modprobe conf file... Anyways I checked the OBS code and it seems that the name "OBS Virtual Camera" is never used elsewhere but in the function creating the v4l2loopback conf.).
* The last one is my physical camera.

The problem seems that the code is checking the fact the `v4l2loopback` module is loaded, but I have no idea how it determines which device to actually use... I 'm guessing the first one.

I tried to use any of the virtual cameras to check if it was not outputing to the wrong virtual camera, but no... **It just reports the virtual cam is started and issues no error, while it obviously doesn't work**.

<hr>
I am runnning Ubuntu 18.04 on this machine
OBS 26.1.0 (64)

FYI, with the previous OBS release I was using the [obs-v4l2sink](https://github.com/CatxFish/obs-v4l2sink) plugin and it was fully working (I removed everything regarding this plugin when I upgraded to this version of OBS).

I hope this helps as it's clearly a blocking issue, I am considering falling back, but would definitely prefer to have OBS virtual camera working with no plugin.

I think **OBS should have a config flag to specify which virtual camera to use (it was the case with the plugin)**. Be it by path, by id, or by whatever is secondary. Even if it was painful to each time check if the `/dev/videox` used was the correct one after a reboot, it was not blocking at all.

**Comment by mattboy9921** [comment]
Using Arch - Rolling, OBS 26.1.0.r20. Can confirm, removing Droidcam and rebooting fixed my issue and Virtual Camera is working beautifully now.

**Comment by lbriais** [comment]
I confirm that **removing all virtual cameras is a workaround not a fix**.
In my case I had to remove both `akvcam` and `droidcam` modules and the the OBS virtual camera started working. But this is no fix !

**The real fix would be to provide a way to specify which v4l2 camera to use as virtual camera like it was the case with the plugin it replaces.**

**Comment by lbriais** [comment]
A much better workaround until this is fixed is to actually continue using the [obs-v4l2sink](https://github.com/CatxFish/obs-v4l2sink) plugins which still allows selecting the relevant virtual camera to use.

Just tested with everything loaded and it works...

**Comment by WizardCM** [comment]
Out of curiosity, how do akvcam and droidcam handle the mapping? Is there something we could be doing better automatically?

**Comment by lbriais** [comment]
I am afraid that, for that, you will have to have a look at their respective repos...
I am just a user in this case.

The point is both akvcam and droidcam use their own module (`akvcam` for Webcamoid and `v4l2loopback_dc` for Droidcam) which are somehow derived from `v4l2loopback` but with at the name differing... I guess this is what they use to target the correct devices.

In the case of OBS, as it relies on the generic `v4l2loopback`, I doubt something fully automated could be really achieved unless maybe only one device exists using that module. The best way is clearly to allow to specify which one to use in the OBS config, and to improve regarding the mechanism currently existing within the plugin, use something else than the `/dev/videox` to improve consistency across reboots...

**Comment by pokgak** [comment]
I'm currently using OBS compiled from source (OBS Studio - 26.1.0-15-g60fed63d6) on elementary (Ubuntu 18.04 base) with the droidcam plugin but not the `v4l2loopback_dc`. Compiled `v4l2loopback` myself and in `/etc/modprobe.d` I have a file `v4l2loopback.conf` with the following content:

```
options v4l2loopback devices=1 video_nr=10 card_label="OBS Virtualcam" exclusive_caps=1
```

Works perfectly for me.

**Comment by JohnDeeBDD** [comment]
I'm having the same issue. For me at least, it's that I can't select WHICH device to output to. My DSLR is a loopback device, and OBS 26 is overwriting it.

Previously, I could select which v4l2loopback device to output to. After updating to OBS 26, with the new "Start Virtual Camera", I can only output to the first v4l2loopback device. Is there a setting somewhere? How can I select the output path of the virtual camera?

Details:
I'm using a Canon DSLR to USB3 via g2photo & v4l2loopback. This is itself a v4l2loopback device. When I run "v4l2-ctl --list-devices", it lists two devices. I then use g2photo to send a stream from the camera to /dev/video0 . For OBS to detect the camera, it has to be running into a loopback device BEFORE OBS starts. The issue is that since I can't select where to send OBS virtual camera output path, it is overwriting the camera with the output. This is how I output OBS into something like Zoom. It was working fine with OBS 25.

**Comment by InFerYes** [comment]
I have the same issue in 26.1. I'm using Droidcam-OBS, the remote camera works in OBS and I can stream/record with it. Clicking `Start Virtual Camera` only produces "Failed to start virtual camera" in output log.

**Comment by daankata** [comment]
Same issue here, again with ArchLinux and DroidCam installed producing "Failed to start virtual camera" in the output.

**Comment by quantenProjects** [comment]
For me, moving droidcam away from `/dev/video0` to another device (eg. `/dev/video3`) somehow worked (Virtual Cam image is somehow disturbed :thinking:):

`/etc/modprobe.d/droidcam.conf`
```
options v4l2loopback_dc width=640 height=480 video_nr=3
```

But `obs-v4l2sink-git` still works better.

Archlinux, OBS 26.1.0-2

**Comment by darrellenns** [comment]
Definitely agree with @lbriais here. An option to specify the loopback device is the best solution. The workaround of moving other applications away from the first loopback device doesn't work in all cases. For example, you can't run multiple OBS instances each with their own loopback output.

**Comment by hollowaykeanho** [comment]
Providing some findings from my side (Debian, build from source with automation dated to master branch 81a89e6) if it is useful. A device selector is suffice.

My v4l2 device is `/dev/video100`.

1. If I use CatFish plugin, I got to select that device - everything is working fine.

![screenshot-2021-02-19-20-42-21](https://user-images.githubusercontent.com/7259572/108505901-09aec480-72f3-11eb-9a00-0740be53da33.png)


2. If I start virtual camera, apparently it locked itself to `/dev/video0` which is my laptop camera (log generated from terminal):
````
...
info: v4l2-input: Start capture from /dev/video0
info: v4l2-input: Input: 0
info: v4l2-input: Resolution: 640x480
info: v4l2-input: Pixelformat: YUYV
info: v4l2-input: Linesize: 1280 Bytes
info: v4l2-input: Framerate: 30.00 fps
info: v4l2-input: /dev/video0: select timeout set to 166666us (5x frame periods)
...
error: v4l2-input: /dev/video0: select timed out
error: v4l2-input: /dev/video0: failed to log status
error: v4l2-input: /dev/video0: select timed out
error: v4l2-input: /dev/video0: failed to log status
error: v4l2-input: /dev/video0: select timed out
error: v4l2-input: /dev/video0: failed to log status
warning: Failed to start virtual camera
```

**Comment by JCGoran** [comment]
Below are some observations I've made.
For reference, here's my setup:

* Debian Unstable
* OBS `26.1.2` without any plugins
* `v4l2loopback-dkms 0.12.5-1` 
* laptop with 1 webcam (which for some reason is both on `/dev/video0` and `/dev/video1`)
* DSLR camera (streamed via `gphoto2`)

### Observations

* if the `v4l2loopback` kernel module isn't loaded, OBS will load it by running `modprobe v4l2loopback exclusive_caps=1 card_label='OBS Virtual Camera'`. On the other hand, if the module is loaded, it won't do anything, such as trying to reload it. This is because dynamic module management isn't [fully implemented yet](https://github.com/umlaeute/v4l2loopback/milestone/1) in `v4l2loopback`, and reloading the whole module would probably break running devices.
* there has to exist at least one free loopback device, otherwise the virtual camera will falsely report that it has started, but won't be available in other software
* on that note, if I create 2 devices, `/dev/video2` and `/dev/video3`, set my DSLR to stream to `/dev/video2`, and try to start the virtual camera, it will produce some bizarre flicker on the output of the DSLR and I won't be able to use the output from the virtual camera in other software. On the other hand, if I stream to `/dev/video3` and start the virtual cam, the virtual camera works fine in other programs (checked with `ffplay /dev/video2` and Signal desktop). **TL;DR** when streaming from devices, it seems that the order matters, and OBS isn't happy if the (numerically) first loopback device isn't available, so make sure that the very first loopback device remains free.
* for any `v4l2loopback` device number > 63, OBS will fail to start the virtual camera; this is [hardcoded](https://github.com/obsproject/obs-studio/blob/ad628eb2e08fc1213d1d405ce4ce36c1c10b3af9/plugins/linux-v4l2/v4l2-output.c#L8), so unless your setup requires it, use device numbers <= 63. Alternatively, you can change the hardcoded constant and recompile OBS from source.
* if all of the above works correctly, and the virtual camera is started, for some reason, I am unable to stream from the webcam anymore (`ffplay /dev/video0` reports `Device or resource busy`), but can stream from both OBS and the DSLR
* if I run OBS from the terminal, occasionally I'll get the message `libv4l2: error dequeuing buf: Invalid argument`; I'm still not sure why that happens, or what triggers it

### Enabling the OBS virtual camera

So, to recap, the following will _probably_ work:

1. when loading the `v4l2loopback` module:
    * always create one extra device for OBS
    * use the option `exclusive_caps=1` (for multiple devices, use `exclusive_caps=1,1,1...`, i.e. put as many ones as there are devices being created)
    * don't use device numbers > 63
2. always leave the numerically first loopback device free, because OBS will use that one (it goes through them [in order](https://github.com/obsproject/obs-studio/blob/ad628eb2e08fc1213d1d405ce4ce36c1c10b3af9/plugins/linux-v4l2/v4l2-output.c#L155))

I will admit that this method is very clumsy, and running multiple OBS instances won't work, but I don't think there's much that can be done, at least until dynamic device management becomes available in `v4l2loopback`, which would enable OBS to spawn a loopback device on-the-fly and use that one.

**Comment by darrellenns** [comment]
I would propose simply adding a new configuration option which allows the user to specify the v4l2loopback device they wish to use. This should allow for all of the more advanced use cases. The user can load the v4l2loopback module with whatever options they want prior to launching OBS (OBS should not attempt to load the module itself if it's already loaded). They can also specify which device they want OBS to use, which eliminates conflicts with other devices/applications and also allows for multiple instances of OBS (by setting different loopback devices for each instance).

If there is a desire to improve automatic behaviour, that can be done in separate patches without affecting users who choose to use the manual setting. Power users will likely always need a manual option for more complex scenarios, regardless of what changes are made to the automatic behaviour.

In my opinion, automatic loading of kernel modules should not be done at all by user applications. It would be better to just give a warning/error message if the module is not loaded, and leave it up to the distribution or user to load the module with the appropriate options. However, I recognize that some may disagree. If we had manual settings, then the user at least has the ability to override the automatic behaviours when they are not correct for their environment.

**Comment by lbriais** [comment]
Back 3 [months ago](https://github.com/obsproject/obs-studio/issues/3929#issuecomment-751330300). Is there any issue with implementing this or are our assumptions wrong somewhere ?
I fully agree on the split of the two problems, one being the automatic default behavior and the other being the manual configuration. 

And if needs to be phased, the manual config solves all the issues whereas the automatic default behavior is just nice to have.

**Comment by sam9032** [comment]
Hello together,
a manual selection would be perfect.
Had the same problem with droidcam and OBS. Both want to use /dev/video0 and either in OBS nor in droidcam, a manual configuration is possible.
I needed a while to figure this out, this thread here helps. Thanks guys! :rocket: 

To solve my problem, I use the solution @quantenProjects describes in his comment and edited /etc/modprobe.d/droidcam.conf from video_nr=0 to video_nr=1.

```
# /etc/modprobe.d/droidcam.conf
options v4l2loopback_dc width=640 height=480 video_nr=1

```
After a reboot, everything is fine. OBS on /dev/video0 and droidcam on /dev/video1.

**Comment by kodwyp** [comment]
Sorry, I have the same problem of OBS having a "Start Virtual Camera" button that doesn't do anything except produce an error message (if OBS is run from command line) of "warning: Failed to start virtual camera".

I am not quite a Linux noob but still need some hand-holding to try to follow the potential solutions described in this thread.  In particular, when _pardomi commented on Dec 21, 2020:
	"Removing v4l2loopback and then manually assigning the next available number (in my case /dev/video2) resulting in the virtual camera working."_
How exactly does one manually assign the next available number?  Is it through the use of the "video_nr=<number>" option?  As in:
	`sudo modprobe v4l2loopback devices=1 video_nr=2 card_label="OBS Virtual Camera" exclusive_caps=1`

If that is the case, then this solution does not work for me.  I have a built-in webcam in my laptop that results in the following output from the command "v4l2-ctl --list-devices":

**Integrated_Webcam_HD: Integrate (usb-0000:00:14.0-11):
        /dev/video0
        /dev/video1**


These devices are present upon boot-up, so I cannot start OBS before them to reserve /dev/video0 for OBS's virtual camera.  When I type
	`sudo modprobe v4l2loopback devices=1 video_nr=2 card_label="OBS Virtual Camera" exclusive_caps=1`
that results in the following output from the command "v4l2-ctl --list-devices":

**OBS Virtual Camera (platform:v4l2loopback-000):
        /dev/video2**

**Integrated_Webcam_HD: Integrate (usb-0000:00:14.0-11):
        /dev/video0
        /dev/video1**

Unfortunately, the error continues to persist, with OBS having a "Start Virtual Camera" button that doesn't do anything when clicked.

I also tried following the suggestion when _lbriais commented on Dec 25, 2020: " I confirm that removing all virtual cameras is a workaround (clearly not a fix)"_

I am not sure whether the built-in webcam which takes up /dev/video0 and /dev/video1 count as virtual cameras.  If they do, then I have no hope: they are already present long before my system can boot up to the point where I can type a "sudo modprobe" command.  If they don't count as virtual cameras, then, for me, removing all virtual cameras is NOT a workaround.  That is, even if I do "sudo modprobe -r v4l2loopback" (which successfully results in "v4l2-ctl --list-devices" NOT listing anything other than the /dev/video[01] above) does NOT solve the problem: when I click on "Start Virtual Camera", OBS asks for sudo permission (I guess to run "sudo modproble v4l2loopback") but still fails to produce a working virtual camera.

I also read where _JCGoran commented on Mar 23: "the following will probably work: ...  when loading the v4l2loopback module: ...  always create one extra device for OBS ... use the option exclusive_caps=1 ..."_ etc.  Unfortunately, I still get the same error.

(The frustrating thing is, up till two weeks ago it had been working for several months, and then one day it simply didn't work.  In between, the only thing that happened was that I installed more RAM on my laptop, but I'm pretty sure that wouldn't cause this error.)

Any insight would be appreciated, but my formal specific question is: How exactly does one manually assign the next available number for the webcam?  And does the presence of the integrated webcam keep me from using OBS's virtual camera?

If this is not the appropriate forum to ask these questions, please forgive me, and let me know where my question would be better posted.

Thanks.

**Comment by mikeperalta1** [comment]
> I think **OBS should have a config flag to specify which virtual camera to use (it was the case with the plugin)**. Be it by path, by id, or by whatever is secondary. Even if with the plugin, it was painful to each time check if the `/dev/videox` used was the correct one after a reboot, it was not blocking at all, which it is now...

+1 to this idea. Config option is the way.

**Comment by DanielRios549** [comment]
I have experienced this using Iriun Webcam. /dev/video0 is created by Iriun on its installation, after that, OBS use this device for its VirtualCam and does not create another one. First, OBS should always create a new device when the user starts it for the first time, even if already exists 50 devices, create a new one is always necessary to avoid conflicts, after that, an option to change is simply unnecessary since OBS already created a new device for it.

**Comment by darrellenns** [comment]
@DanielRios549 The loopback devices are created when the v4l2loopback kernel module is loaded. There is no provision for applications to create new ones after the module has already been loaded. However, the kernel module parameters devices= and video_nr= can be used to create whatever number and naming of loopback devices is desired when the module is loaded.

The obvious thing OBS should be doing is just allowing users to select the desired loopback device, rather than naively assuming that they will only ever use the first one. Even if dynamically adding loopbacks was possible (which it currently is not), it would still be necessary in many cases to specify a specific device for consistent naming purposes.

**Comment by mikeperalta1** [comment]
> @DanielRios549 The loopback devices are created when the v4l2loopback kernel module is loaded. There is no provision for applications to create new ones after the module has already been loaded. 

Would it be possible/practical to scan all existing loopback devices, unload the module, then reload the module with all previous devices *plus* a new one? Obviously that might not be desired behavior for everyone, but a button with a warning+confirmation could be nice, for users who aren't great at the CLI themselves.

Then would of course be a separate feature, somewhere near the new feature where we select a device.

**Comment by WPettersson** [comment]
> Would it be possible/practical to scan all existing loopback devices, unload the module, then reload the module with all previous devices _plus_ a new one? Obviously that might not be desired behavior for everyone, but a button with a warning+confirmation could be nice, for users who aren't great at the CLI themselves.

If I was going to start putting that much effort into this, I would choose to improve the v4l loopback module code so that one can dynamically request a new V4L loopback device. That feels less clunky, less likely to break things (i.e. for those that have multiple devices, allows one to keep working when you load a second), and would be easier to test + maintain.

**Comment by darrellenns** [comment]
@mikeperalta1
> Would it be possible/practical to scan all existing loopback devices, unload the module, then reload the module with all previous devices plus a new one?

If any other applications have a v4l2loopback device open, the module unload would just fail. Also, this would not solve the problem of consistent device naming, so you would have to be sure to start things in the same order or end up reconfiguring your applications each time. If people truly want dynamic add/remove of loopback modules, that capability would have to be added to v4l2loopback (along with some type of persistent device naming).

I would propose a much simpler solution. Add a configuration option to allow users to set which loopback device they want to use. If it's not specified, keep the current behavior of using the first device. For users with the simple scenario (one OBS instance and no other loopback devices), everything is automatic just like it is now. For users with more complex scenarios, they can just specify the device they want to use.

**Comment by koopdi** [comment]
There appears to be some support for dynamic device management, although it's not available in my distro yet. That could solve this issue(s).[1]
Additionally, choosing the virtual devices based on attributes other than their sequence number would probably be a good idea.[2] e.g. prefer the first loopback device with a label containing the string "OBS" and avoid loopback devices that /do/ report sink capabilities (they are likely already in use by another source)
[1]https://github.com/umlaeute/v4l2loopback#dynamic-device-management
[2]https://wiki.archlinux.org/title/Udev#udev_rule_example

**Comment by 6ffm70** [comment]
An observation that may be related:
I'm using a laptop with an integrated webcam (which appears to be registering as four different devices), 
```
Integrated Camera: Integrated C (usb-0000:00:14.0-8):
        /dev/video0
        /dev/video1
        /dev/video2
        /dev/video3 
```
and a dSLR which I use as a webcam via v4l2/ffmpeg/gphoto, which I had assigned _video_nr=5 card_label="dSLR"_
```
dSLR (platform:v4l2loopback-000):
        /dev/video5
```
Both cameras work fine when opened in any application. They also work fine when opened in OBS, until I start the virtual camera; then the image of the dSLR gets distorted. I use Chromium&Jitsi to test the output of the OBS virtual cam.

![image](https://user-images.githubusercontent.com/50849255/120572716-ba614f80-c41c-11eb-9e5f-1c375ffe6ce5.png)
Preview in OBS before starting virtual camera

![image](https://user-images.githubusercontent.com/50849255/120572776-d4029700-c41c-11eb-8885-1768fc849531.png)
Preview in OBS after starting virtual camera

![image](https://user-images.githubusercontent.com/50849255/120572997-2f348980-c41d-11eb-992b-14552795bc2b.png)
Resulting image in Jitsi (the OBS virtual camera is labelled 'dSLR' in the Jitsi camera selection)

However, when using the internal webcam (video0) via OBS, the image will not get distorted when used in Jitsi.

**Comment by JCGoran** [comment]
> There appears to be some support for dynamic device management, although it's not available in my distro yet. That could solve this issue(s).[1]
> Additionally, choosing the virtual devices based on attributes other than their sequence number would probably be a good idea.[2] e.g. prefer the first loopback device with a label containing the string "OBS" and avoid loopback devices that /do/ report sink capabilities (they are likely already in use by another source)
> [1]https://github.com/umlaeute/v4l2loopback#dynamic-device-management
> [2]https://wiki.archlinux.org/title/Udev#udev_rule_example

Unfortunately, the device management hasn't been implemented yet upstream, see [this comment](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=985507#10) from the maintainer on the Debian bug tracker, which references issue https://github.com/umlaeute/v4l2loopback/issues/398.

**Comment by zakpatterson** [comment]
I was noticing the virtual camera would work one time, then no UI response and `Failed to start virtual camera` in the logs after the first time until reboot. I discovered just using `modprobe -r` would fix it for me.

So if I see `Failed to start virtual camera`, my workaround is:
```
➜  ~ sudo modprobe -r v4l2loopback
modprobe: FATAL: Module v4l2loopback is in use.
```
So exit `obs`
```
➜  ~ v4l2-ctl --list-devices      
OBS Virtual Camera (platform:v4l2loopback-000):
	/dev/video10
Lenovo EasyCamera: Lenovo EasyC (usb-0000:00:14.0-6):
	/dev/video0
	/dev/video1
```
Note even after exiting, the virtual camera still exists. Get rid of that.
```
➜  ~ sudo modprobe -r v4l2loopback
➜  ~ v4l2-ctl --list-devices      
Lenovo EasyCamera: Lenovo EasyC (usb-0000:00:14.0-6):
	/dev/video0
	/dev/video1
```
Now restart `obs`. 
```
➜  ~ sudo modprobe v4l2loopback exclusive_caps=1
➜  ~ v4l2-ctl --list-devices                    
OBS Virtualcam (platform:v4l2loopback-000):
	/dev/video10

Lenovo EasyCamera: Lenovo EasyC (usb-0000:00:14.0-6):
	/dev/video0
	/dev/video1
```
Now virtual camera works.

**Comment by TheK0tYaRa** [comment]
Basically,
`pkexec sh -c 'pamac install droidcam --no-confirm' ; droidcam ; pkexec sh -c 'pamac remove droidcam --no-confirm ; pamac reinstall v4l2loopback-dkms --no-confirm ; modprobe v4l2loopback'`

**Comment by TheK0tYaRa** [comment]
o no

**Comment by insou22** [comment]
+1, just ran into this same issue today on Ubuntu LTS with droidcam before finding this issue.

**Comment by arcans** [comment]
**THIS IS NOT AN ISSUE FROM OBS | MAY 2022**
>Normally, OBS uses default increment to add virtual camera.

**When you run this BEFORE RUNNING ANY VIRTUAL CAM/DEVICES:**
`$v4l2-ctl --list-devices`
**Normally, It will show you this:**
```
Integrated Camera: Integrated C (usb-0000:00:14.0-8):
/dev/video0
 ```
**BUT, when you run the above command AFTER RUNNING OBS VIRTUAL CAMERA:**
```
Integrated Camera: Integrated C (usb-0000:00:14.0-8):
/dev/video0
/dev/video1
```
**To see what /dev/video1 is playing, run this:**
`$ffplay /dev/video1`
>A window will pop up playing yourvideo1.mp4

**Plus, when you run the above command AFTER RUNNING ANOTHER VIRTUAL CAMERA(NOT AN OBS VIRTUAL CAMERA) THAT USES /dev/video1:**
```
Integrated Camera: Integrated C (usb-0000:00:14.0-8):
/dev/video0
/dev/video1
```
**To see what /dev/video1 is playing once again, run this:**
`$ffplay /dev/video1`
>A window will pop up playing yourvideo2.mp4

**SOLUTION 1**:
>> You can request a feature to use a highly customizable and dynamic videoX where X is a new number you want to assign

**SOLUTION 2 (Recommended as of today May 2022)**:
>> You can use the terminal alone to make virtual camera and you don't need to reboot your system after making any changes
>> Make sure you run as root@... in your terminal.
>> Run rmmod v4l2loopback to remove previously loaded v4l2loopback kernel/module
>> Use [v4l2loopback official repo](https://github.com/umlaeute/v4l2loopback)  for creating a new Virtual Camera with a new name and a new videoX of your own.
>> and ffmpeg for flushing the video output.

**Comment by darrellenns** [comment]
@arcans Those solutions do not address situations where you have multiple applications (other OBS instances or other applications) and multiple v4l2loopback devices. There needs to be a way to tell OBS which loopback device it should use. Otherwise, there will inevitably be conflicts and unpredictable behaviour. The current automatic device selection make work for the most basic use case, but I strongly believe that there needs to be a setting for power users to be able to choose which loopback device to use.

**Comment by hollowaykeanho** [comment]
Just tested the latest e513d1cb6b337ff0cb13c7aa79b3026421c48ad5 on secure-boot Debian Bullseye OS. Screenshot is my latest finding (where VLC is showing a severe frame lag).

![screenshot-2022-05-06-07-08-01](https://user-images.githubusercontent.com/7259572/167041520-2c2e6940-353d-473c-9701-34feefb0d260.png)

It seems that:
1. OBS somehow already has a way to walk through each `/dev/videoX` device until it found the right one (thanks?). There weren't any changes to my kernel module load out (still `/dev/video100`) as custom device.
2. I do face some severe frame lag (~10s) with the built-in virtual camera (I suspect is due to the 4k frame size so further investigation will be done). Log:

```
info: Output 'virtualcam_output': Total frames output: 5654
info: Output 'virtualcam_output': Total drawn frames: 5133 (11858 attempted)
info: Output 'virtualcam_output': Number of lagged frames due to rendering lag/stalls: 6725 (56.7%)
```


Can anyone else re-produce the same results please?

-----

Strictly speaking by UNIX convention and design:

1. OBS dev team is **NOT SUPPOSEDLY** meddle with how `v4l2loopback` kernel module is being loaded. What they can do is only up until advising the general way of loading and that's it. The rest are the Linux users' and its power users' responsibilities.
1.1. There are infinite way Linux OS can be configured left alone `v4l2loopback`. It's next to impossible to test all permutations and combinations, left alone covering all of them.

2. OBS is a user-app. All it needs it is just to offer a method for selecting a video device for loopback offered by the kernel and that's it. The bonus can probably be persisting the last dialog value inside config file.
2.1. I know automated scanning is seamlessly good UX but for a tool dealing with "real-time" application, precise control is very important especially dealing with I/O devices.
2.2. PERSONAL OPINION: status-quo of auto-scanning is acceptable as long as the maintainers can manage the workload and preferred this way.

----

@arcans, I disagree with solution 1 because this ticket does not need to include a kernel loading feature as a simple query dialog for which device to use is suffice. Hence, the problem is still within OBS domain. The existing dev-team already implemented an automatic scanner.

I agree with your `v4l2loopback` point (see my points on "OBS vs Kernel" above).

---- 

Peace. Not here to incite quarrel. Just helping out to make a great software greater. =)

**Comment by japrogramer** [comment]
Im using obs 27.2.4-1 on Arch btw and when I create my loopback device ffmpeg works fine with it but obs does not see it,
Is the automatic /dev/videoX scanner not on this release?

```

➜  ~ sudo modprobe v4l2loopback exclude_caps=1 card_label=VirtualVideoDevice
➜  ~ v4l2-ctl --list-devices
VirtualVideoDevice (platform:v4l2loopback-000):
        /dev/video2

HP Wide Vision HD Camera: HP Wi (usb-0000:03:00.3-3):
        /dev/video0
        /dev/video1
        /dev/media0
```

**Comment by ickedude** [comment]
I am using OBS to do some filtering before using the stream in Microsoft Teams. My Pipeline looks like this:
/dev/video0 -> python opencv -> /dev/video12 -> obs -> /dev/video11 -> teams

I loaded the v4l2loopback module with the option below and had the same problems as described in this issue.
modprobe v4l2loopback devices=2 exclusive_caps=1 video_nr=11,12 card_label=obs,py

After some weeks I came back to this topic and by coincidence created the devices the other way around and the issue was gone.
modprobe v4l2loopback devices=2 exclusive_caps=1 video_nr=**12,11** card_label=py,obs

Maybe someone can validate if it is a usefull workaround.

**Comment by MrCybertux** [comment]
I do not know if my Problem is related but i get the following message from the logs:
```
10:11:53.177: NV12 texture support not available
10:11:53.177: P010 texture support not available
10:11:53.183: Failed to start virtual camera
```
There is also strange behavior around the install from my apt repository's and the Flatpack
After Starting the Flatpack will not show any virtual cam button.
The Apt installed version will and up on clicking i get a promt to enter a sudoers password.
After doing that in the apt version Flatpack suddenly shows the button for the virtual cam.
Both do not display anything.
Worked like a charm a few updates ago.
The Log data:
[obslog.txt](https://github.com/obsproject/obs-studio/files/10848633/obslog.txt)

**Comment by heyakyra** [comment]
> 
> **SOLUTION 2 (Recommended as of today May 2022)**:
> 
> > > You can use the terminal alone to make virtual camera and you don't need to reboot your system after making any changes
> > > **1. Make sure you run as root@... in your terminal.**
> > > `$sudo su -`
> > > **2. (Required) To remove previously loaded v4l2loopback kernel/module. Note: You can only load v4l2loopback once so we need to unload it but not uninstall**
> > > `$rmmod v4l2loopback`
> > > **Step 3. Run v4l2loopback as root**
> > > `sudo modprobe v4l2loopback`
> > > **4. Create a new virtual camera named Your Cam Label here with custom path /dev/video69**
> > > `modprobe v4l2loopback video_nr=69 card_label="Your Cam Label here"`
> > > **5. Use ffmpeg for flushing the best video output.**
> > > `$ffmpeg -stream_loop -1 -i yourorig.mp4 -c copy tempvideofile.mp4 `
> > > If you appreciate my answer, consider donate https://www.paypal.com/paypalme/cacansoft?country.x=PH&locale.x=en_US

Sadly this doesn't appear to work on Fedora Silverblue due to https://github.com/coreos/rpm-ostree/issues/1091#issuecomment-683469742

Is there another workaround that would be compatible?

**Comment by Cyanic** [comment]
Foung a solution here: https://obsproject.com/forum/threads/huge-amount-of-error-logs-when-camera-doesnt-return-framerate.145191/

**Comment by mhagnumdw** [comment]
For me, on Fedora 40, the virtual camera worked fine. After a system update it stopped working. After a long time trying to solve it and only seeing the message `warning: Failed to start virtual camera` in the OBS log without any other details, using `sudo lsof | grep '/dev/video'` I noticed that `/dev/video0` was busy, **then I discovered that stopping the `v4l2-relayd` service (`sudo systemctl stop v4l2-relayd`) made the OBS virtual camera work again!**

For now I don't know what v4l2-relayd is for or if I need it.

My devices before and after start virtual camera:
```
$ v4l2-ctl --list-devices
Intel MIPI Camera (platform:v4l2loopback-000):
	/dev/video0
```

**Comment by stephematician** [comment]
For those following, here's what _OBS_ does on Linux (in a nutshell) - I hope this helps identify workarounds:

-   On startup, _OBS_ checks if the _v4l2loopback_ kernel module is available.
-   Once you select `Start Virtual Camera`:
    -   _OBS_ checks if the module is _loaded_.
    -   If the _loopback_ module is not loaded, _OBS_ will try the command `modprobe v4l2loopback exclusive_caps=1 card_label='OBS Virtual Camera'` (this is when it asks for your password).
        -   The _loopback_ module will create one node at `/dev/video#` where `#` is the first unused integer (starting at 0).
    -    _OBS_ searches through the files that start with `/dev/video` and _grabs the first one that can set an output format_. It searches in alpha-numerical order, e.g. `/dev/video0`, `/dev/video1`, ….

Crucially, _OBS_ does _not_ check if a node was actually spawned by _v4l2loopback_. It can output to __any__ V4L2 (device) with output capabilities.

As noted in #12081, it doesn't even check that it was able to select the correct format and size - hence, I believe, what @6ffm70 sees: https://github.com/obsproject/obs-studio/issues/3929#issuecomment-853497569. PR #12083 should apply better constraints on this.

It looks like the developers intended for _OBS_ to _only use v4l2loopback devices_, but maybe this isn't the best option given that there are a number of loopback-like modules available on Linux (droidcam, akvcam, etc).

If there is scope to loosen that requirement, then I would support the user having control over which node is selected - along with a useful default (e.g. the current behaviour), along with an option to (attempt to) create a node with the _v4l2loopback_ module if no output device is found.


### How OBS checks if the kernel module is available

First it checks if it is _loaded_: It reads `/proc/modules` and searches for __any__ substring match to `v4l2loopback` - so it _can_ match to things like _v4l2loopback_dc_ too - i.e. it can generate false positives.

If not loaded, it checks if it is _available_: It tests for errors running `modinfo v4l2loopback`, which is at least more specific.

The check on whether the module is loaded could (possibly) be replaced by looking at `/dev/v4l2loopback` instead, which has been available for some time.

**Comment by Humam-Hamdan** [comment]
When i try to start virtual camera obs goes into 0mb ram usage and i can't use the camera until after a reboot, this is the case for both the obs in my distro's repos (32.0.2 @ OpenMandirva ROME) and the flatpak version (32.1.2).

**Comment by stephematician** [comment]
@Humam-Hamdan - that seems more like a user support request? Have you tried the forums: https://obsproject.com/forum/list/linux-support.34/
